plugins/inputs/firehose/firehose_request.go

```diff
@@ -0,0 +1,156 @@
+package firehose
+
+import (
+	"compress/gzip"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/influxdata/telegraf/config"
+)
+
+type firehoseRecord struct {
+	EncodedData string `json:"data"`
+}
+
+type firehoseRequestBody struct {
+	RequestID string           `json:"requestId"`
+	Timestamp int64            `json:"timestamp"`
+	Records   []firehoseRecord `json:"records"`
+}
+
+// https://docs.aws.amazon.com/firehose/latest/dev/httpdeliveryrequestresponse.html#requestformat
+type firehoseRequest struct {
+	req                *http.Request
+	body               firehoseRequestBody
+	responseStatusCode int
+}
+
+func (r *firehoseRequest) authenticate(expectedAccessKey config.Secret) error {
+	if expectedAccessKey.Empty() {
+		return nil
+	}
+	accessKey, err := expectedAccessKey.Get()
+	if err != nil {
+		return fmt.Errorf("getting accesskey failed: %w", err)
+	}
+	reqAccessKey := r.req.Header.Get("x-amz-firehose-access-key")
+	if reqAccessKey != accessKey.String() {
+		r.responseStatusCode = http.StatusUnauthorized
+		return fmt.Errorf("unauthorized request %s", r.req.Header.Get("x-amz-firehose-request-id"))
```

Comment:
Please log the source of the request (aka remote address) if possible.

Reply:
done